### ISSUE TITLE
Pylance parameter hints not working with child class attribute

### ISSUE THREAD

**Issue by mozTheFuzz** [microsoft/pylance-release#3512]
## Issue

I am working on dataclass objects inherent faust.Record [faust-streaming library](https://github.com/faust-streaming/faust)

Somehow pylance parameter hints will ignore the class attribute and goes to the parent __init__ method for its attributes.

The default parameter hints work just fine until I turn pylance on.

I tried to create a class with __inti__ parameters and pylance occasionally breaks.

Is there any system preferences I should be aware of? 
Thank you. 

## Environment data

-   Language Server version: 2022.10.30
-   OS and version: MasOS 12.4 arm64
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9.7

## Code Snippet

```python

from dataclasses import dataclass
import faust


@dataclass
class B(faust.Record):
    color: str


B()
```

## Repro Steps

Have a dataclass object to inherent from faust.Record.

## Expected behavior

![no_pylance](https://user-images.githubusercontent.com/77670573/197160076-133d384e-f6b3-4b92-8b24-a07675f25eb1.gif)

## Actual behavior

![pylance](https://user-images.githubusercontent.com/77670573/197160094-815582a7-9232-4df8-a046-93cc7d5d4954.gif)

## Logs

```
(21976) Background analysis message: getDiagnosticsForRange
(21976) Background analysis message: getDiagnosticsForRange
(21976) Background analysis message: setFileOpened
(21976) Background analysis message: markFilesDirty
(21976) [FG] completion at /Users/my-mac/Documents/Coding/Python/booking/playground.py:9:1 ...
(21976) [FG]   parsing: /Users/my-mac/Documents/Coding/Python/booking/playground.py (0ms)
(21976) [FG]   binding: /Users/my-mac/Documents/Coding/Python/booking/playground.py (0ms)
(21976) [FG] completion at /Users/my-mac/Documents/Coding/Python/booking/playground.py:9:1 [found 427 items] (18ms)
(21976) IntelliCode: current invocation did not produce any meaningful tokens.
(21976) Time taken to get recommendations: 2 ms, Memory increase: -542 KB.
(21976) Background analysis message: getDiagnosticsForRange
(21976) Background analysis message: getDiagnosticsForRange
(21976) Background analysis message: getSemanticTokens delta
(21976) [BG(1)] getSemanticTokens delta previousResultId:1666344142120 at /Users/my-mac/Documents/Coding/Python/booking/playground.py ...
(21976) [BG(1)]   parsing: /Users/my-mac/Documents/Coding/Python/booking/playground.py (0ms)
(21976) [BG(1)]   binding: /Users/my-mac/Documents/Coding/Python/booking/playground.py (1ms)
(21976) [BG(1)] getSemanticTokens delta previousResultId:1666344142120 at /Users/my-mac/Documents/Coding/Python/booking/playground.py (5ms)
(21976) Background analysis message: analyze
(21976) [BG(1)] analyzing: /Users/my-mac/Documents/Coding/Python/booking/playground.py ...
(21976) [BG(1)]   checking: /Users/my-mac/Documents/Coding/Python/booking/playground.py (1ms)
(21976) [BG(1)] analyzing: /Users/my-mac/Documents/Coding/Python/booking/playground.py (1ms)
(21976) Background analysis message: resumeAnalysis
(21976) [BG(1)] indexing: /Users/my-mac/Documents/Coding/Python/booking/playground.py [found 1] (0ms)
(21976) Indexing Done: /Users/my-mac/Documents/Coding/Python/booking/playground.py
(21976) Background analysis message: setFileOpened
(21976) Background analysis message: markFilesDirty
(21976) [FG] parsing: /Users/my-mac/Documents/Coding/Python/booking/playground.py (0ms)
(21976) [FG] binding: /Users/my-mac/Documents/Coding/Python/booking/playground.py (0ms)
(21976) Background analysis message: analyze
(21976) [BG(1)] analyzing: /Users/my-mac/Documents/Coding/Python/booking/playground.py ...
(21976) [BG(1)]   parsing: /Users/my-mac/Documents/Coding/Python/booking/playground.py (1ms)
(21976) [BG(1)]   binding: /Users/my-mac/Documents/Coding/Python/booking/playground.py (0ms)
(21976) [BG(1)]   checking: /Users/my-mac/Documents/Coding/Python/booking/playground.py (6ms)
(21976) [BG(1)] analyzing: /Users/my-mac/Documents/Coding/Python/booking/playground.py (7ms)
(21976) Background analysis message: getDiagnosticsForRange
(21976) Background analysis message: resumeAnalysis
(21976) [BG(1)] indexing: /Users/my-mac/Documents/Coding/Python/booking/playground.py [found 1] (0ms)
(21976) Indexing Done: /Users/my-mac/Documents/Coding/Python/booking/playground.py
(21976) Background analysis message: getDiagnosticsForRange
(21976) Background analysis message: getSemanticTokens delta
(21976) [BG(1)] getSemanticTokens delta previousResultId:1666344146071 at /Users/my-mac/Documents/Coding/Python/booking/playground.py (0ms)
```

**Comment by debonte** [comment]
To determine the parameters of `__init__` on a dataclass we need to not only look for fields in the current class, but also in any dataclasses in its class hierarchy. The problem here is that there are some unknown types in the class hierarchy of `B`.

The class hierarchy is `B` -> `Record` -> `Model` -> `ModelT` -> `base`. Where `base` is defined as: 

```python
base = abc.ABC if abc_compatible_with_init_subclass else object
```

Since we can't determine statically what type `base` is, we can't reliably determine the parameter list of `__init__` and therefore add default `args` and `kwargs` parameters to avoid false positive errors.

The behavior that you are seeing is intentional. I'd suggest that you file an issue on Faust. One possible solution would be for them to provide type checkers with a simpler value for `base`. For example, maybe this:

```python
if TYPE_CHECKING
    base = abc.ABC
else
    base = abc.ABC if abc_compatible_with_init_subclass else object
```

**Comment by mozTheFuzz** [comment]
Hi @debonte  Thank you for helping but I am still a bit confused....

The parameters hints in my case are:

```python
( self, *args: Any, __strict__: bool = True, __faust: Any = None, **kwargs: Any ) -> None:
```

The above hints seemed to derive from the __init__ function of Record under faust.models.records.py line 311

```python

def __init__(
        self, *args: Any, __strict__: bool = True, __faust: Any = None, **kwargs: Any
    ) -> None:  # pragma: no cover
        ...  # overridden by _BUILD_init
```

It does not seem to got deeper than the class Record.

**Comment by mozTheFuzz** [comment]
@debonte I did try to include type hinting and it works now.


```python
base:Union[abc.ABC,object] = abc.ABC if abc_compatible_with_init_subclass else object
```

My understanding is that pylanc looks into the hierarchy of "B" and goes all the way down to "base".
Since base is not statically typed, it goes back down to its hierarchy agent to find the __init__ function and return its parameters for suggestion?

**Comment by debonte** [comment]
> I did try to include type hinting and it works now.
> 
> ```python
> base:Union[abc.ABC,object] = abc.ABC if abc_compatible_with_init_subclass else object
> ```

This is not correct. `base: Union[abc.ABC, object]` indicates that `base` is an instance of either `abc.ABC` or `object`, but it's neither of those things -- it's a type.

And if you hover over `base` you'll see that Pylance already understands that. There's no need to provide a type annotation on `base`.

![image](https://user-images.githubusercontent.com/12820898/197948082-89382d2e-de21-4a2a-b4ad-a7d8b5083c6d.png)

> The parameters hints in my case are:
> 
> ```python
> ( self, *args: Any, __strict__: bool = True, __faust: Any = None, **kwargs: Any ) -> None:
> ```

I agree, that is surprising. I'll look into this.

I still believe that my initial explanation is correct though. And you should open an issue on Faust suggesting the `if TYPE_CHECKING` change I mentioned above.

**Comment by debonte** [comment]
> The above hints seemed to derive from the init function of Record under faust.models.records.py line 311

@mozTheFuzz, thanks for pointing this out. This has been fixed in Pyright and will be included in Pylance's next prerelease build.

**Comment by rchiodo** [comment]
This issue has been fixed in prerelease version 2022.11.21, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20221121-9-November-2022-PreRelease)
